internal/error.go

```diff
@@ -369,3 +479,260 @@ func newUnknownExternalWorkflowExecutionError() *UnknownExternalWorkflowExecutio
+func IsRetryable(err error, nonRetryableTypes []string) bool {
+	var terminatedErr *TerminatedError
+	var canceledErr *CanceledError
+	var workflowPanicErr *workflowPanicError
+	if errors.As(err, &terminatedErr) || errors.As(err, &canceledErr) || errors.As(err, &workflowPanicErr) {
+		return false
+	}
+
+	var applicationErr *ApplicationError
+	var applicationErrOriginalType string
+	if errors.As(err, &applicationErr) {
+		if applicationErr.nonRetryable {
+			return false
+		}
+		applicationErrOriginalType = applicationErr.orginalType
+	}
+
+	var timeoutErr *TimeoutError
+	if errors.As(err, &timeoutErr) {
+		if timeoutErr.timeoutType != commonpb.TimeoutType_StartToClose &&
+			timeoutErr.timeoutType != commonpb.TimeoutType_Heartbeat {
+			return false
+		}
+	}
+
+	var serverErr *ServerError
+	if errors.As(err, &serverErr) {
+		if serverErr.nonRetryable {
+			return false
+		}
+	}
+
+	for {
+		causeErr := errors.Unwrap(err)
+		if causeErr == nil {
+			break
+		}
+		err = causeErr
+	}
+	errType := getErrorType(err)
+	for _, nonRetryableType := range nonRetryableTypes {
+		if nonRetryableType == errType || nonRetryableType == applicationErrOriginalType {
+			return false
+		}
+	}
+
+	return true
+}
+
+func getErrorType(err error) string {
+	var t reflect.Type
+	for t = reflect.TypeOf(err); t.Kind() == reflect.Ptr; t = t.Elem() {
+	}
+
+	return t.Name()
+}
+
+// convertErrorToFailure converts error to failure.
+func convertErrorToFailure(err error, dc DataConverter) *failurepb.Failure {
+	if err == nil {
+		return nil
+	}
+
+	failure := &failurepb.Failure{
+		Source:  "GoSDK",
+		Message: err.Error(),
+	}
+
+	switch err := err.(type) {
+	case *ApplicationError:
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type:         getErrorType(err),
+			NonRetryable: err.nonRetryable,
+			Details:      convertErrDetailsToPayloads(err.details, dc),
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+	case *CanceledError:
+		failureInfo := &failurepb.CanceledFailureInfo{
+			Details: convertErrDetailsToPayloads(err.details, dc),
+		}
+		failure.FailureInfo = &failurepb.Failure_CanceledFailureInfo{CanceledFailureInfo: failureInfo}
+	case *PanicError:
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type: getErrorType(err),
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+		failure.StackTrace = err.StackTrace()
+	case *workflowPanicError:
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type:         getErrorType(&PanicError{}),
+			NonRetryable: true,
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+		failure.StackTrace = err.StackTrace()
+	case *TimeoutError:
+		failureInfo := &failurepb.TimeoutFailureInfo{
+			TimeoutType:          err.timeoutType,
+			LastFailure:          convertErrorToFailure(err.lastErr, dc),
```

Comment:
Can you confirm where lastErr is coming from?  Is this on the Timeout history event?  It would be great if we can add an integration test for Timeout error with lastFailure and heartbeat details.

Reply:
It comes from `ActivityTaskTimedOutEventAttributes.LastFailure` inside `handleActivityTaskTimedOut` handler. This `LastFailure` is set on server, apparently, to the last failure from history. This logic hasn't changed in this PR but so far I don't see that it's getting set correctly. This is actually the part of my next task (and another PR).